util/pingpong.c

```diff
@@ -0,0 +1,2111 @@
+	float usec_per_xfer;
+
+	if (sent == 0)
+		return;
+
+	if (name) {
+		if (header) {
+			printf("%-50s%-8s%-8s%-9s%-8s%8s %10s%13s%13s\n",
+					"name", "bytes", "#sent", "#ack",
+					"total", "time", "MB/sec",
+					"usec/xfer", "Mxfers/sec");
+			header = 0;
+		}
+
+		printf("%-50s", name);
+	} else {
+		if (header) {
+			printf("%-8s%-8s%-9s%-8s%8s %10s%13s%13s\n",
+					"bytes", "#sent", "#ack", "total",
+					"time", "MB/sec", "usec/xfer",
+					"Mxfers/sec");
+			header = 0;
+		}
+	}
+
+	printf("%-8s", size_str(str, tsize));
+
+	printf("%-8s", cnt_str(str, sent));
+
+	if (sent == acked) {
+		printf("=%-8s", cnt_str(str, acked));
+	} else if (sent < acked) {
+		printf("-%-8s", cnt_str(str, abs(acked - sent)));
+	} else {
+		printf("+%-8s", cnt_str(str, abs(acked - sent)));
+	}
+
+	printf("%-8s", size_str(str, bytes));
+
+	usec_per_xfer = ((float)elapsed / sent / xfers_per_iter);
+	printf("%8.2fs%10.2f%11.2f%11.2f\n",
+		elapsed / 1000000.0, bytes / (1.0 * elapsed),
+		usec_per_xfer, 1.0/usec_per_xfer);
+}
+
+void ft_pingpong_usage(char *name, char *desc)
+{
+	fprintf(stderr, "Usage:\n");
+	fprintf(stderr, "  %s [OPTIONS]\t\tstart server\n", name);
+	fprintf(stderr, "  %s [OPTIONS] <host>\tconnect to server\n", name);
+
+	if (desc)
+		fprintf(stderr, "\n%s\n", desc);
+
+	fprintf(stderr, "\nOptions:\n");
+	FT_PRINT_OPTS_USAGE("-b <src_port>", "non default source port number");
+	FT_PRINT_OPTS_USAGE("-p <dst_port>", "non default destination port number");
+	FT_PRINT_OPTS_USAGE("-s <address>", "source address");
+
+	FT_PRINT_OPTS_USAGE("-n <domain>", "domain name");
+	FT_PRINT_OPTS_USAGE("-f <provider>", "specific provider name eg sockets, verbs");
+	FT_PRINT_OPTS_USAGE("-e <ep_type>", "Endpoint type: msg|rdm|dgram (default:dgram)");
+
+	FT_PRINT_OPTS_USAGE("-I <number>", "number of iterations");
+	FT_PRINT_OPTS_USAGE("-S <size>", "specific transfer size or 'all'");
+
+	FT_PRINT_OPTS_USAGE("-h", "display this help output");
+
+	FT_PRINT_OPTS_USAGE("-v", "enables data_integrity checks");
+
+	return;
+}
+
+void ft_parse_opts(struct ct_pingpong *ct, int op, char *optarg)
+{
+	switch (op) {
+	
+	/* Domain */
+	case 'n':
+		if (!ct->hints->domain_attr) {
+			ct->hints->domain_attr = malloc(sizeof *(ct->hints->domain_attr));
+			if (!ct->hints->domain_attr) {
+				perror("malloc");
+				exit(EXIT_FAILURE);
+			}
+		}
+		ct->hints->domain_attr->name = strdup(optarg);
+		break;
+	
+	/* Fabric */
+	case 'f':
+		if (!ct->hints->fabric_attr) {
+			ct->hints->fabric_attr = malloc(sizeof *(ct->hints->fabric_attr));
+			if (!ct->hints->fabric_attr) {
+				perror("malloc");
+				exit(EXIT_FAILURE);
+			}
+		}
+		ct->hints->fabric_attr->prov_name = strdup(optarg);
+		break;
+	
+	/* Endpoint */
+	case 'e':
+		if (!strncasecmp("msg", optarg, 3))
+			ct->hints->ep_attr->type = FI_EP_MSG;
+		if (!strncasecmp("rdm", optarg, 3))
+			ct->hints->ep_attr->type = FI_EP_RDM;
+		if (!strncasecmp("dgram", optarg, 5))
+			ct->hints->ep_attr->type = FI_EP_DGRAM;
```

Comment:
Add a base case here to handle BS input. So if I do:

```sh
bturrubi at savbu-usnic-a :: ./fi_pingpong -e adsfasdf -d
[debug] util/pingpong.c:418: SERVER: waiting for connection ...
```

That should have failed.